src/client.rs

```diff
@@ -1019,64 +1016,75 @@ impl Client {
 
     /// Compute an RSASSA-PKCS#1v1.5 signature of the SHA-256 hash of the given data.
     ///
-    /// **WARNING**: This functionality has not been tested and has not yet been
-    /// confirmed to actually work! USE AT YOUR OWN RISK!
-    ///
     /// You will need to enable the `untested` cargo feature to use it.
     ///
     /// <https://developers.yubico.com/YubiHSM2/Commands/Sign_Pkcs1.html>
-    #[cfg(feature = "untested")]
-    pub fn sign_rsa_pkcs1v15_sha256(
+    pub(crate) fn sign_rsa_pkcs1v15<S: SignatureAlgorithm>(
         &self,
         key_id: object::Id,
         data: &[u8],
     ) -> Result<rsa::pkcs1::Signature, Error> {
         Ok(self
             .send_command(SignPkcs1Command {
                 key_id,
-                digest: Sha256::digest(data).as_slice().into(),
+                digest: S::digest(data).as_slice().into(),
             })?
             .into())
     }
 
-    /// Compute an RSASSA-PSS signature of the SHA-256 hash of the given data with the given key ID.
-    ///
-    /// **WARNING**: This functionality has not been tested and has not yet been
-    /// confirmed to actually work! USE AT YOUR OWN RISK!
+    /// Compute an RSASSA-PKCS#1v1.5 signature of the SHA-256 hash of the given data.
     ///
     /// You will need to enable the `untested` cargo feature to use it.
     ///
+    /// <https://developers.yubico.com/YubiHSM2/Commands/Sign_Pkcs1.html>
+    pub fn sign_rsa_pkcs1v15_sha256(
+        &self,
+        key_id: object::Id,
+        data: &[u8],
+    ) -> Result<rsa::pkcs1::Signature, Error> {
+        self.sign_rsa_pkcs1v15::<Sha256>(key_id, data)
+    }
+
+    /// Compute an RSASSA-PSS signature of the SHA-256 hash of the given data with the given key ID.
+    ///
     /// <https://developers.yubico.com/YubiHSM2/Commands/Sign_Pss.html>
-    #[cfg(feature = "untested")]
-    pub fn sign_rsa_pss_sha256(
+    pub(crate) fn sign_rsa_pss<S: SignatureAlgorithm>(
         &self,
         key_id: object::Id,
         data: &[u8],
     ) -> Result<rsa::pss::Signature, Error> {
         ensure!(
-            data.len() > rsa::pss::MAX_MESSAGE_SIZE,
+            data.len() < rsa::pss::MAX_MESSAGE_SIZE,
             ErrorKind::ProtocolError,
             "message too large to be signed (max: {})",
             rsa::pss::MAX_MESSAGE_SIZE
         );
 
-        let mut hasher = Sha256::default();
-
-        let length = data.len() as u16;
-        hasher.update(length.to_be_bytes());
```

Comment:
to note: I dropped this. I couldn't find where it was coming from, but that was producing incorrect signature.

Reply:
(and this is a breaking change, but ... this is fixing signatures?)